editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,583 @@
+					rect = rect.grow(-1);
+					draw_rect(rect, border_col, false, 1);
+				} else {
+					border_col.a = 0.9;
+					draw_rect(rect, border_col, false, 1);
+					draw_line(Vector2(handle_x_pos, 0), Vector2(handle_x_pos, h / 2 - 1), border_col);
+					if (hovered_index == i) {
+						// Draw a subtle translucent rect inside the handle if it's being hovered.
+						rect = rect.grow(-1);
+						border_col.a = 0.54;
+						draw_rect(rect, border_col, false, 1);
+					}
+				}
+			}
+
+			// Draw "button" for color selector.
+			int button_offset = total_w + handle_width + draw_spacing;
+			if (selected_index != -1) {
+				Color grabbed_col = gradient->get_color(selected_index);
+				if (grabbed_col.a < 1) {
+					draw_texture_rect(get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons")), Rect2(button_offset, 0, h, h), true);
+				}
+				draw_rect(Rect2(button_offset, 0, h, h), grabbed_col);
+				if (grabbed_col.r > 1 || grabbed_col.g > 1 || grabbed_col.b > 1) {
+					// Draw an indicator to denote that the currently selected color is "overbright".
+					draw_texture(get_theme_icon(SNAME("overbright_indicator"), SNAME("ColorPicker")), Point2(button_offset, 0));
+				}
 			} else {
-				draw_texture_rect(icon, Rect2(margin, margin, icon->get_width(), icon->get_height()));
+				// If no color is selected, draw grey color with 'X' on top.
+				draw_rect(Rect2(button_offset, 0, h, h), Color(0.5, 0.5, 0.5, 1));
+				draw_line(Vector2(button_offset, 0), Vector2(button_offset + h, h), Color(0.8, 0.8, 0.8));
+				draw_line(Vector2(button_offset, h), Vector2(button_offset + h, 0), Color(0.8, 0.8, 0.8));
+			}
+		} break;
+
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (!is_visible()) {
+				grabbing = GRAB_NONE;
 			}
 		} break;
 	}
 }
 
-Size2 GradientReverseButton::get_minimum_size() const {
-	return (get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons"))->get_size() + Size2(margin * 2, margin * 2));
+void GradientEditorRect::_on_mouse_exited() {
+	hovered_index = -1;
+	queue_redraw();
+}
+
+void GradientEditorRect::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_selected_index", "index"), &GradientEditorRect::set_selected_index);
+	ClassDB::bind_method(D_METHOD("reverse_gradient"), &GradientEditorRect::reverse_gradient);
+}
+
+GradientEditorRect::GradientEditorRect() {
+	set_focus_mode(FOCUS_ALL);
+
+	popup = memnew(PopupPanel);
+	picker = memnew(ColorPicker);
+	popup->add_child(picker);
+	popup->connect("about_to_popup", callable_mp(EditorNode::get_singleton(), &EditorNode::setup_color_picker).bind(GradientEditorRect::get_picker()));
+
+	preview_texture.instantiate();
+	preview_texture->set_width(1024);
+	add_child(popup, false, INTERNAL_MODE_FRONT);
+}
+
+///////////////////////
+
+void GradientEditor::_reverse_button_pressed() {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Reverse Gradient"));
```

Comment:
```suggestion
	undo_redo->create_action(TTR("Reverse Gradient"), UndoRedo::MERGE_DISABLE, gradient.ptr());
```
Required to properly use scene local history.

Reply:
Hmm, something is telling me this might result in problems, but let me try.

Reply:
Well, GradientEditor is not the one who holds the gradient. So I put this in another `_reverse_gradient` func in GradientEditorRect, please take a look to make sure it's right as tbh I don't understand the problem

Reply:
Looks good now. Previously if you had built-in Gradient on a scene, reverse would register as global history action instead of scene.